Revamp ITR documentation (#17286)
DataDog/documentation: content/en/continuous_integration/intelligent_test_runner/_index.md

    tag: "Blog"
    text: "Monitor all your CI pipelines with Datadog"
---

> <div class="alert alert-warning">Intelligent Test runner is a beta product.</div>

Reviewer: ```suggestion
<div class="alert alert-warning">Intelligent Test Runner is a beta product.</div>
```

> Intelligent Test Runner is Datadog's test impact analysis solution. It automatically selects and runs only the relevant tests needed to validate any given commit, enabling you to reduce testing downtime while maintaining test coverage.

Author: The text in this page I've basically copied from https://www.datadoghq.com/blog/streamline-ci-testing-with-datadog-intelligent-test-runner/

Reviewer 2: "Test Impact Analysis" is the industry term for what we're doing, the analysis of which tests are impacted. I think it's helpful to include in terms of SEO (Search engine optimization). Datadog docs pages rank quite high with google, so hopefully when someone searches for "Test impact analysis" our docs page can show up.

I also agree that "testing downtime" is unclear/misleading.

What do you think about this combination of suggestions:

"Intelligent Test Runner is Datadog's test impact analysis solution. It automatically and intelligently selects and runs only the relevant tests for a given commit based on the code being changed. Significantly reduce time spent testing and overall CI costs, while maintaining test coverage."

> Intelligent Test Runner is Datadog's test impact analysis solution. It automatically and intelligently selects and runs only the relevant tests for a given commit based on the code being changed. Significantly reduce time spent testing and overall CI costs, while maintaining test coverage.

Reviewer 3: "automatically and intelligently" sounds more like marking material than docs. That's just my opinion though

Reviewer 3: I'm sure the docs team will come up with something better but I'd rather do something like

> It uses the code changed in a given commit to automatically select and run only the relevant tests.

```suggestion
Intelligent Test Runner is Datadog's test impact analysis solution. It uses the code changed in a given commit to automatically select and run only the relevant tests. It significantly reduces time spent testing and overall CI costs, while maintaining test coverage.
```

Author: I see this initial text is contentious! Haha! There's already a thread about this in https://github.com/DataDog/documentation/pull/17286#discussion_r1135944042

@KickingTheTV, I think you have the last word here.

{{< img src="continuous_integration/itr_savings.png" alt="Intelligent test runner enabled in a test session showing its time savings.">}}

It does this by analyzing your test suite to determine the code each test covers, and then cross-referencing that coverage with the files impacted by a new code change. Datadog uses this information to run a selection of relevant, impacted tests, omitting the ones unaffected by the code change and reducing the overall testing duration.

By minimizing the number of tests run per commit, Intelligent Test Runner reduces the frequency of flaky tests disrupting your pipelines. Flaky tests are tests that may pass or fail at random given the same commit. This can be particularly frustrating when the test flaking is unrelated to the code change being tested. After enabling Intelligent Test Runner for your test services, you can limit each commit to its relevant tests to ensure that flaky tests unrelated to your code change don’t end up arbitrarily breaking your build.

> ### Limitations during beta

Reviewer 3: is that a limitation only during the beta? We might want to set expectations here

There are known limitations in the current implementation of Intelligent Test Runner that can cause it to skip tests that should be run under certain conditions. Specifically, Intelligent Test Runner is not able to detect changes in library dependencies, compiler options, external services or changes to data files in data-driven tests.

To override Intelligent Test Runner and run all tests, add `ITR:NoSkip` (case insensitive) anywhere in your Git commit message.

> ## Setup Datadog Library

Reviewer 4: ```suggestion
## Set up Datadog library
```
Let's take the opportunity to bring this heading into compliance with our capitalization and wording standards.

> Prior to setting up Intelligent Test Runner, you must have finished setting up [Test Visibility][1] for your particular language. If you are reporting data through the Datadog Agent you need to use v6.40+/v7.40+.

Reviewer 4: ```suggestion
Prior to setting up Intelligent Test Runner, you must set up [Test Visibility][1] for your particular language. If you are reporting data through the Agent, use v6.40+/v7.40+.
```

{{< whatsnext desc="Choose a language to set up Intelligent Test Runner in Datadog:" >}}
    {{< nextlink href="continuous_integration/intelligent_test_runner/dotnet" >}}.NET{{< /nextlink >}}
    {{< nextlink href="continuous_integration/intelligent_test_runner/javascript" >}}JavaScript{{< /nextlink >}}
    {{< nextlink href="continuous_integration/intelligent_test_runner/swift" >}}Swift{{< /nextlink >}}
{{< /whatsnext >}}

## Configuration

> Once you have setup your Datadog Library for Intelligent Test Runner, you can configure it from the [Test Service Settings][2] page.

Reviewer 4: ```suggestion
Once you have setup your Datadog library for Intelligent Test Runner, configure it from the [Test Service Settings][2] page.
```

{{< img src="continuous_integration/itr_overview.png" alt="Intelligent test runner enabled in test service settings in the CI section of Datadog." style="width:80%;">}}

> Due to the limitations described above, the default branch of your repository is automatically excluded from having Intelligent Test Runner enabled. This is the recommended configuration by Datadog which ensures that all of your tests are run prior to reaching production.

Reviewer 4: ```suggestion
Due to the limitations described above, the default branch of your repository is automatically excluded from having Intelligent Test Runner enabled. Datadog recommends this configuration to ensure that all of your tests run prior to reaching production.
```

> Due to the limitations described above, the default branch of your repository is automatically excluded from having Intelligent Test Runner enabled. This is the recommended configuration by Datadog to ensure that all of your tests are run in the default branch.

Reviewer 3: the last phrase does not explain much, does it? 
Maybe 
```suggestion
Due to the limitations described above, the default branch of your repository is automatically excluded from having Intelligent Test Runner enabled. This is the recommended configuration by Datadog to minimize the risks of false positives. 
```

Author: Discussed offline and rephrased.

> If there are other branches you want to exclude, you can add them from the Intelligent Test Runner settings page. The query bar supports the wildcard character `*` to exclude any branches that match.

Reviewer 3: maybe add an example? 
```suggestion
If there are other branches you want to exclude, you can add them from the Intelligent Test Runner settings page. The query bar supports the wildcard character `*` to exclude any branches that match, such as `release_*`
```

> If there are other branches you want to exclude, you can add them from the Intelligent Test Runner settings page. The query bar supports the wildcard character `*` to exclude any branches that match, such as `release_*`

Reviewer 4: ```suggestion
If there are other branches you want to exclude, add them from the Intelligent Test Runner settings page. The query bar supports using the wildcard character `*` to exclude any branches that match, such as `release_*`
```

{{< img src="continuous_integration/itr_configuration.png" alt="Select branches to exclude from intelligent test runner" style="width:80%;">}}

## Explore test sessions

> You can explore the time savings you are getting from Intelligent Test Runner by looking at the test commit page and test sessions panel.

Reviewer 4: ```suggestion
You can explore the time savings you get from Intelligent Test Runner by looking at the test commit page and test sessions panel.
```

{{< img src="continuous_integration/itr_commit.png" alt="Test commit page with intelligent test runner" style="width:80%;">}}

{{< img src="continuous_integration/itr_savings.png" alt="Intelligent test runner enabled in a test session showing its time savings." style="width:80%;">}}

> When Intelligent Test Runner is active and skipping tests, you can see a purple text which displays the amount of time saved on each test session or on each commit. The duration bar also changes color to purple so you can quickly identify which test sessions are using Intelligent Test Runner in the [Test Runs][3] page.

Reviewer 4: ```suggestion
When Intelligent Test Runner is active and skipping tests, purple text displays the amount of time saved on each test session or on each commit. The duration bar also changes color to purple so you can quickly identify which test sessions are using Intelligent Test Runner in the [Test Runs][3] page.
```

## Explore adoption and global savings

> You can also track you organization's savings and adoption of Intelligent Test Runner through our out-of-the-box [Intelligent Test Runner dashboard][4], which includes widgets to track your overall savings as well as a per-repository, per-committer and per-service view of the data so you can understand where and who is getting the most out of Intelligent Test Runner.

Reviewer 2: I would change this last sentence to something more like:

"...so you can understand what parts of your organization are using and getting the most out of Intelligent Test Runner."

> You can also track you organization's savings and adoption of Intelligent Test Runner through the out-of-the-box [Intelligent Test Runner dashboard][4]. The dashboard includes widgets to track your overall savings as well as a per-repository, per-committer and per-service view of the data so you can understand what parts of your organization are using and getting the most out of Intelligent Test Runner.

Reviewer 4: ```suggestion
Track you organization's savings and adoption of Intelligent Test Runner through the out-of-the-box [Intelligent Test Runner dashboard][4]. The dashboard includes widgets to track your overall savings as well as a per-repository, per-committer, and per-service view of the data. View the dashboard to understand which parts of your organization are using and getting the most out of Intelligent Test Runner.
```

{{< img src="continuous_integration/itr_dashboard1.png" alt="Intelligent Test Runner dashboard" style="width:80%;">}}

> It also allows you to track adoption of Intelligent Test Runner throughout your organization.

Reviewer 4: ```suggestion
The dashboard also tracks adoption of Intelligent Test Runner throughout your organization.
```

{{< img src="continuous_integration/itr_dashboard2.png" alt="Intelligent Test Runner dashboard" style="width:80%;">}}

## Further Reading

{{< partial name="whats-next/whats-next.html" >}}

[1]: /continuous_integration/tests/
[2]: https://app.datadoghq.com/ci/settings/test-service
[3]: https://app.datadoghq.com//ci/test-runs

> [4]: https://app.datadoghq.com/dash/integration/30941/ci-visibility-intelligent-test-runner-beta

Author: Need to confirm this link with @drodriguezhdez
